File: Problems/hacker_rank/20141222_Sherlock_and_GCD/Solution.java
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class Solution  {
  public static boolean hasSubset(final int[] arr)  {
    int i = 2;
    int maxNum = 0;
    for ( Integer a : arr ) {
      if ( maxNum < a )
        maxNum = a;
    }
    Set<Integer> noSet = new HashSet<Integer>();
    noSet.add(0);

    Set<Integer> primeSet = new HashSet<Integer>();
    while ( i < maxNum + 1 )  {
      primeSet.clear();
      for ( Integer a : arr ) {
        primeSet.add(a % i);
      }
      if ( primeSet.equals(noSet) )
        return false;
      ++i;
    }
    return true;
  }
  public static void main(String[] args)  {
    Scanner sc = new Scanner(System.in);
    int T = Integer.parseInt(sc.nextLine());
    for ( int i = 0; i < T; ++i ) {
      int N = Integer.parseInt(sc.nextLine());
      String[] inp = sc.nextLine().split(" ");
      int[] arr = new int[inp.length];
      for ( int j = 0; j < inp.length; ++j )  {
        arr[j] = Integer.parseInt(inp[j]);
      }
      if ( hasSubset(arr) )
        System.out.println("YES");
      else
        System.out.println("NO");
    }
  }
}

File: Problems/hacker_rank/20141222_Sherlock_and_GCD/Solution.scala
object Solution {
  def hasSubset(arr: Array[Int]): Boolean = {
    var i = 2
    val maxNum = arr.reduceLeft(_ max _)
    val noSet = Set(0)
    while ( i < maxNum + 1 )  {
      if ( arr.map(_ % i).toSet == noSet )  {
        return false
      }
      i += 1
    }
    true
  }
  def main(args: Array[String]) {
    val inps = io.Source.stdin.getLines.toArray.drop(1)
    for ( i <- 0 until inps.length by 2 ) {
      val arr = inps(i + 1).split(" ").map((s: String) => s.toInt)
      if ( hasSubset(arr) ) {
        println("YES")
      } else  {
        println("NO")
      }
    }
  }
}
